upstairs/src/lib.rs

```diff
@@ -9097,49 +9115,39 @@ impl Guest {
         let mut bw_check_ok = true;
         let mut iop_check_ok = true;
 
-        // XXX if recv ever is called from multiple threads, token locks must be
-        // taken for the whole of the procedure, not multiple times in the below
-        // if blocks!
```

Comment:
This PR looks correct from an async-cancel-safety perspective. I'm less sure about the semantic (or performance, maybe?) change affected by acquiring all three locks up front, where the code previously only acquired the later two locks if and when needed (and kept them locked for a shorter duration). Will defer to you and Alan.

Reply:
Looking back, I don't think the previous code was correct if consume_req was called from multiple tasks / threads (it's currently not, hence removing that comment). There definitely seems to have been a TOCTOU problem.